src/port.rs

```
@@ -0,0 +1,211 @@
+        } else if self.ports[port].state != CspPortStateT::PortClosed {
+            Err(CspPortError::PortAlreadyBound)
+        } else {
+            self.ports[port].state = CspPortStateT::PortOpen;
+            self.ports[port].socket = socket;
+            Ok(())
+        };
+
+        result
+    }
+
+    pub fn close(&mut self, port: usize) -> Result<(), CspPortError> {
+        let result = if port >= CSP_PORT_MAX_BIND {
+            Err(CspPortError::CspInvalidBindPort)
+        } else if self.ports[port].state == CspPortStateT::PortClosed {
+            Err(CspPortError::PortAlreadyClosed)
+        } else {
+            self.ports[port].state = CspPortStateT::PortClosed;
+            self.ports[port].socket = CspSocketT::default();
+            Ok(())
+        };
+        result
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::basic_types::CspServicePortT;
+    use crate::port::CSP_PORT_MAX_BIND;
+    use crate::port::{CspPortStateT, CspSocketT, Ports};
+    use crate::port::{CspQueueHandleT, CspStaticQueueT};
+
+    #[test]
+    fn test_csp_socket_new() {
+        let mut ports = Ports::new();
+
+        // create & test default socket structure
+        let socket = CspSocketT::default();
+        assert_eq!(socket.opts, 0);
+        assert_eq!(socket.rx_queue_static_data, '\0');
+        assert_eq!(socket.rx_queue_static, CspStaticQueueT::default());
+        assert_eq!(socket.rx_queue, CspQueueHandleT::default());
+
+        // bind socket to a port
+        // TODO: we are casting CspServicePortT::CspAny to usize as a hack here, but this is not great.
+        // I think we should switch to using bit flags for the ports. Much like it is in csp_header_flags!
+        // Even more so because we do not use socket references, we might want to
+        // bind a socket to any combination of ports, which would be easy to do with bit flags.
+        let result = ports.bind(socket, CspServicePortT::CspAny as usize);
```

@TimMeiwald Please take a look at this, i'm pretty sure there's a more practical way to do this.

---

Why do we need to case to USize can the bind function not just take CspServicePortT directly? But yes if we can have multiple values in one usize then we may need to use the bitfields. 

If however we are already diverging a little we can just use some struct instead with booleans for each thing? We can then just use some bitfield flag thing to cast from the public API into our own struct since using a few more bytes during runtime temporarily to bind a socket is imo fine and maybe less confusing?

---

"_Why do we need to case to USize can the bind function not just take CspServicePortT directly?_" - No, because we use it as an index in `bind()`.

---

Using it as an index is fine but we can cast to usize inside of bind so that if it's an invalid value we can return an appropriate error(which you already do anyway)?

If multiple options can be passed with one usize then yyeah we need some bitflags thing or a struct of options and then a way to read a usize to cast it to that struct of options. The latter is probably a rustier option because whilst temporarily using more stack memory it saves on computation compared to the bitflags and doesn't appear to actually get stored anywhere(yet anyway). 

If only one option can be chosen at a time I'd just cast inside the function instead. Requiring the user to pass a usize makes it more likely for them to pass an incorrect value.

---

I'll cast inside the function. This still means that a single socket can be bound to either one of the available ports or all of them. This is an obvious limitation that we'd want to get rid of in the future. For now, we leave it like this, but I'll create an issue to address this.

---

This is done, too.